File: programmers/level3/단어_변환.java
class Solution {
	int answer = Integer.MAX_VALUE;
	String t;
    public int solution(String begin, String target, String[] words) {
        t = target;
        boolean[] visited = new boolean[words.length];
        dfs(begin, words,visited, 0);
        if(answer == Integer.MAX_VALUE) answer = 0;
        return answer;
    }
    void dfs(String current, String[] words, boolean[] visited, int count) {
    	if(current.equals(t)) {
    		if(count < answer) {
    			answer = count;
    		}
    		return;
    	}
    	for(int i = 0; i < visited.length; ++i) {
    		if(!visited[i] && countDiff(current,words[i])==1) {
    			visited[i] = true;
    			dfs(words[i], words, visited, count+1);
    			visited[i] = false;
    		}
    	}
    }
    int countDiff(String a, String b) {
    	int count = 0;
    	for(int i = 0; i < a.length(); ++i) {
    		if(a.charAt(i)!=b.charAt(i)) {
    			count++;
    		}
    	}
    	return count;
    }
}
public class 단어_변환 {
	public static void main(String[] args) {
		String begin = "hit";
		String target = "cog";
		String[] words = {"hot", "dot", "dog", "lot", "log", "cog"};
		Solution sol = new  Solution();
		System.out.println(sol.solution(begin, target, words));
	}
}

File: programmers/js/level3/단어_변환.js
var answer = 987654321;
function solution(begin, target, words) {
    let visited = Array.from(Array(words.length), () => { false });
    dfs(0, begin, target, visited, words);
    if (answer == 987654321) answer = 0;
    return answer;
}

const dfs = (count, current, target, visited, words) => {
    if (count > answer) {
        return;
    }
    if (current == target) {
        if (count < answer) {
            answer = count;
        }
        return;
    }
    for (let i = 0; i < visited.length; ++i) {
        const d = countDiff(current, words[i]);
        if (d == 1 && !visited[i]) {
            visited[i] = true;
            dfs(count + 1, words[i], target, visited, words);
            visited[i] = false;
        }
    }
}
const countDiff = (current, word) => {
    let count = 0;
    for (let i = 0; i < word.length; ++i) {
        if (current[i] != word[i]) {
            count++;
        }
    }
    return count;
}

console.log(solution("hit", "cog", ["hot", "dot", 'dog', 'lot', 'log', 'cog']));